crates/steel-parser/src/lexer.rs

```diff
@@ -147,38 +150,47 @@ impl<'a> Lexer<'a> {
 
             Some(&code @ ('x' | 'u')) => {
                 self.eat();
+                let start = self.token_end - 2;
 
                 let mut digits = String::new();
 
-                let braces = match self.chars.peek().copied() {
+                let escape_end = match self.chars.peek().copied() {
                     Some('{') if code == 'u' => {
                         self.eat();
-                        true
+                        '}'
                     }
-                    _ => false,
+                    _ => ';',
                 };
 
-                loop {
+                let valid = loop {
                     let Some(c) = self.eat() else {
-                        return Err(TokenError::MalformedByteEscape);
+                        return Err(incomplete);
                     };
 
                     match c {
-                        ';' if !braces => break,
-                        '}' if braces => break,
-                        c if c.is_ascii_digit() => {
-                            digits.push(c);
-                        }
-                        'a'..='f' | 'A'..='F' => {
-                            digits.push(c);
-                        }
-                        _ => return Err(TokenError::MalformedByteEscape),
+                        c if c == escape_end => break true,
+                        // note that this overlaps partially with `escapeEnd`
+                        ';' | '\\' | '\n' | '(' | ')' | '[' | ']' | '{' | '}' => break false,
+                        c if c == delim => break false,
+                        _ => digits.push(c),
```

Comment:
The logic in this loop has changed significantly. Now we don't fail as early as we find an invalid character. Instead we try to continue as much as we can until we find some kind of "delimiter".

That way we can report unclosed literals. And in the case with invalid characters gets caught anyway by the `u32` parsing.